types/wlr_layer_shell.c

```diff
@@ -0,0 +1,384 @@
+#define _POSIX_C_SOURCE 200809L
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_layer_shell.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "wlr-layer-shell-unstable-v1-protocol.h"
+
+static void resource_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct zwlr_layer_shell_v1_interface layer_shell_impl;
+static struct zwlr_layer_surface_v1_interface layer_surface_impl;
+
+static struct wlr_layer_client *layer_client_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_layer_shell_v1_interface,
+		&layer_shell_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_layer_surface *layer_surface_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_layer_surface_v1_interface,
+		&layer_surface_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void layer_surface_configure_destroy(
+		struct wlr_layer_surface_configure *configure) {
+	if (configure == NULL) {
+		return;
+	}
+	wl_list_remove(&configure->link);
+	free(configure);
+}
+
+static void layer_surface_handle_ack_configure(struct wl_client *client,
+		struct wl_resource *resource, uint32_t serial) {
+	struct wlr_layer_surface *surface = layer_surface_from_resource(resource);
+
+	bool found = false;
+	struct wlr_layer_surface_configure *configure, *tmp;
+	wl_list_for_each_safe(configure, tmp, &surface->configure_list, link) {
+		if (configure->serial < serial) {
+			layer_surface_configure_destroy(configure);
+		} else if (configure->serial == serial) {
+			found = true;
+			break;
+		} else {
+			break;
+		}
+	}
+	if (!found) {
+		wl_resource_post_error(surface->client->resource,
+			ZWLR_LAYER_SURFACE_V1_ERROR_INVALID_SURFACE_STATE,
+			"wrong configure serial: %u", serial);
+		return;
+	}
+
+	surface->configured = true;
+	surface->configure_serial = serial;
+
+	surface->current.anchor = configure->state.anchor;
+	surface->current.exclusive_zone = configure->state.exclusive_zone;
+	surface->current.margin = configure->state.margin;
+
+	layer_surface_configure_destroy(configure);
+}
+
+static void layer_surface_handle_set_anchor(struct wl_client *client,
+		struct wl_resource *resource, uint32_t anchor) {
+	struct wlr_layer_surface *surface = layer_surface_from_resource(resource);
+	surface->next.anchor = anchor;
+}
+
+static void layer_surface_handle_set_exclusive_zone(struct wl_client *client,
+		struct wl_resource *resource, uint32_t zone) {
+	struct wlr_layer_surface *surface = layer_surface_from_resource(resource);
+	surface->next.exclusive_zone = zone;
+}
+
+static void layer_surface_handle_set_margin(struct wl_client *client,
+		struct wl_resource *resource, uint32_t top,
+		uint32_t right, uint32_t bottom, uint32_t left) {
```

Comment:
Should be `int32_t` for negatives.

Reply:
Margins can't be negative

Reply:
We already discussed this in the previous iteration and decided negative margins could be useful for animations.

Reply:
The use case is for guake-style terminals that can use this to animate downward from the top of the screen.